sql: span use-after-Finish in "batch flow coordinator"

https://github.com/cockroachdb/cockroach/pull/75424 adds some more span use-after-Finish detection. With that patch, the `streamproducer` package immediately finds a violation. The span that's being used after having been `Finish()` ed is the `batch flow coordinator` one.

@yuzefovich will you please help me with this? :)


```
COCKROACH_DEBUG_SPAN_USE_AFTER_FINISH=true make test PKG=./pkg/ccl/streamingccl/streamproducer
```

The span was finished at:
```
panic: use of Span after Finish. Span: batch flow coordinator. Finish previously called at: goroutine 13453 [running]:
runtime/debug.Stack()
        /home/andrei/goroot/src/runtime/debug/stack.go:24 +0x65
github.com/cockroachdb/cockroach/pkg/util/tracing.(*Span).finishInternal(0xc0043f3500)
        /home/andrei/src/github.com/cockroachdb/cockroach/pkg/util/tracing/span.go:256 +0x73
github.com/cockroachdb/cockroach/pkg/util/tracing.(*Span).Finish(0x517be20)
        /home/andrei/src/github.com/cockroachdb/cockroach/pkg/util/tracing/span.go:247 +0x19
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*BatchFlowCoordinator).Run(0xc002e36dc0, {0x5208c88, 0xc001289f40})
        /home/andrei/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/flow_coordinator.go:287 +0x477
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*vectorizedFlow).Run(0xc00367cbe0, {0x5208c88, 0xc001289f40}, 0x1)
        /home/andrei/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/vectorized_flow.go:259 +0x1d9
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run(0xc0059deb00, 0xc00271ae10, 0xc0005f0000, 0xc003ddcc80, 0xc002caf500, 0xc000c4cf08, 0x0)
        /home/andrei/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:563 +0xb56
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRun(0x5208d30, {0x5208d30, 0xc004749a70}, 0x1, 0xc00271ae10, 0xc003ab26a8, {{0x520a818, 0xc0043154a0}, 0x0}, 0xc002caf500)
        /home/andrei/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:1401 +0x191
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithDistSQLEngine(0xc000c4c600, {0x5208d30, 0xc004749a70}, 0xc000c4cc48, 0xc004749a70, {0x7fb1f693bdf8, 0xc003ab26a8}, 0x0, 0xc004315498)
        /home/andrei/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1408 +0x5f4
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).dispatchToExecutionEngine(0xc000c4c600, {0x5208d30, 0xc004749a70}, 0xc000c4cc48, {0x7fb1f693bdf8, 0xc003ab26a8})
        /home/andrei/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1085 +0x9c7
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState(0xc000c4c600, {0x5208d30, 0xc004749a70}, {{0x52370f8, 0xc0018c5090}, {0xc0042360bc, 0x34}, 0x2, 0x0}, 0xc003bc1700, ...)
        /home/andrei/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:722 +0x2526
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt(0xc000c4c600, {0x5208d30, 0xc004749a10}, {{0x52370f8, 0xc0018c5090}, {0xc0042360bc, 0x34}, 0x2, 0x0}, 0xc003bc1700, ...)
        /home/andrei/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:135 +0x572
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execPortal(0xc000c4c600, {0x5208d30, 0xc004749a10}, {0xc003bc1700, {0xc002a979c0, 0x2, 0x2}, {0xc0044809ac, 0x1, 0x1}, ...}, ...)
        /home/andrei/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:222 +0x158
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd.func2({{0xc00423614c, 0x0}, 0x0, {0x2657b1f3, 0xed97f7562, 0x0}}, 0xc000c4c600, 0xc0043a1a38, 0xc0043a1a28, 0xc0043a1948, ...)
        /home/andrei/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1820 +0x511
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd(0xc000c4c600, {0x5208c88, 0xc001289d40})
        /home/andrei/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1822 +0x43a
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run(0xc000c4c600, {0x5208c88, 0xc001136ec0}, 0xc0012a14a0, {0xc00375de10, 0x4201fd, 0xc00420b220}, 0xc005881ee0)
        /home/andrei/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1676 +0x285
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn(0xc003ab2000, {0x5208c88, 0xc001136ec0}, {0xc004892300}, {0x3, 0xc001287ca0, 0xc00420b220}, 0xc000854048)
        /home/andrei/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:681 +0x9a
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).processCommandsAsync.func1()
        /home/andrei/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:669 +0x2f5
```

and later used at:

```
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).closeWrapper(0xc000c4c600, {0x5208c88, 0xc001136ec0}, {0x39dcd60, 0xc0051c83d0})
        /home/andrei/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:975 +0x1a5
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func1()
        /home/andrei/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:679 +0x53
panic({0x39dcd60, 0xc0051c83d0})
        /home/andrei/goroot/src/runtime/panic.go:1038 +0x215
github.com/cockroachdb/cockroach/pkg/util/tracing.(*Span).detectUseAfterFinish(0x5208d30)
        /home/andrei/src/github.com/cockroachdb/cockroach/pkg/util/tracing/span.go:183 +0x134
github.com/cockroachdb/cockroach/pkg/util/tracing.(*Span).Tracer(...)
        /home/andrei/src/github.com/cockroachdb/cockroach/pkg/util/tracing/span.go:226
github.com/cockroachdb/cockroach/pkg/util/log.getSpanOrEventLog({0x5208d30, 0xc004749f20})
        /home/andrei/src/github.com/cockroachdb/cockroach/pkg/util/log/trace.go:92 +0x54
github.com/cockroachdb/cockroach/pkg/util/log.logfDepth({0x5208d30, 0xc004749f20}, 0x1, 0x3, 0x0, {0x412c79f, 0x2c}, {0xc00439f948, 0x2, 0x2})
        /home/andrei/src/github.com/cockroachdb/cockroach/pkg/util/log/channels.go:55 +0x2a8
github.com/cockroachdb/cockroach/pkg/util/log.Errorf(...)
        /home/andrei/src/github.com/cockroachdb/cockroach/pkg/util/log/log_channels_generated.go:640
github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamproducer.(*eventStream).Close(0xc001b0c600, {0x5208d30, 0xc004749f20})
        /home/andrei/src/github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamproducer/event_stream.go:207 +0x136
github.com/cockroachdb/cockroach/pkg/sql/rowexec.(*projectSetProcessor).close(0xc004d50a00)
        /home/andrei/src/github.com/cockroachdb/cockroach/pkg/sql/rowexec/project_set.go:291 +0x9a
github.com/cockroachdb/cockroach/pkg/sql/rowexec.(*projectSetProcessor).ConsumerClosed(0xc0043f3858)
        /home/andrei/src/github.com/cockroachdb/cockroach/pkg/sql/rowexec/project_set.go:299 +0x19
github.com/cockroachdb/cockroach/pkg/sql/execinfra.(*ProcessorBaseNoHelper).InternalClose(0xc000701c00)
        /home/andrei/src/github.com/cockroachdb/cockroach/pkg/sql/execinfra/processorsbase.go:911 +0x149
github.com/cockroachdb/cockroach/pkg/sql/colexec.(*Columnarizer).Close(0xc00439fa58)
        /home/andrei/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/columnarizer.go:296 +0x30
github.com/cockroachdb/cockroach/pkg/sql/colexec.(*invariantsChecker).Close(0xc001289f40)
        /home/andrei/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/invariants_checker.go:135 +0x35
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*vectorizedFlowCreator).setupFlow.func1.1.1()
        /home/andrei/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/vectorized_flow.go:1139 +0x58
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*callbackCloser).Close(0xc000ce7900)
        /home/andrei/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/vectorized_flow.go:1047 +0x1a
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*BatchFlowCoordinator).close(0xc002e36dc0)
        /home/andrei/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/flow_coordinator.go:335 +0x8f
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*BatchFlowCoordinator).Run.func1()
        /home/andrei/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/flow_coordinator.go:260 +0x2b
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*BatchFlowCoordinator).Run(0xc002e36dc0, {0x5208c88, 0xc001289f40})
        /home/andrei/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/flow_coordinator.go:287 +0x498
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*vectorizedFlow).Run(0xc00367cbe0, {0x5208c88, 0xc001289f40}, 0x1)
        /home/andrei/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/vectorized_flow.go:259 +0x1d9
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run(0xc0059deb00, 0xc00271ae10, 0xc0005f0000, 0xc003ddcc80, 0xc002caf500, 0xc000c4cf08, 0x0)
        /home/andrei/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:563 +0xb56
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRun(0x5208d30, {0x5208d30, 0xc004749a70}, 0x1, 0xc00271ae10, 0xc003ab26a8, {{0x520a818, 0xc0043154a0}, 0x0}, 0xc002caf500)
        /home/andrei/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:1401 +0x191
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithDistSQLEngine(0xc000c4c600, {0x5208d30, 0xc004749a70}, 0xc000c4cc48, 0xc004749a70, {0x7fb1f693bdf8, 0xc003ab26a8}, 0x0, 0xc004315498)
        /home/andrei/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1408 +0x5f4
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).dispatchToExecutionEngine(0xc000c4c600, {0x5208d30, 0xc004749a70}, 0xc000c4cc48, {0x7fb1f693bdf8, 0xc003ab26a8})
        /home/andrei/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1085 +0x9c7
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState(0xc000c4c600, {0x5208d30, 0xc004749a70}, {{0x52370f8, 0xc0018c5090}, {0xc0042360bc, 0x34}, 0x2, 0x0}, 0xc003bc1700, ...)
        /home/andrei/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:722 +0x2526
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt(0xc000c4c600, {0x5208d30, 0xc004749a10}, {{0x52370f8, 0xc0018c5090}, {0xc0042360bc, 0x34}, 0x2, 0x0}, 0xc003bc1700, ...)
        /home/andrei/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:135 +0x572
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execPortal(0xc000c4c600, {0x5208d30, 0xc004749a10}, {0xc003bc1700, {0xc002a979c0, 0x2, 0x2}, {0xc0044809ac, 0x1, 0x1}, ...}, ...)
        /home/andrei/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:222 +0x158
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd.func2({{0xc00423614c, 0x0}, 0x0, {0x2657b1f3, 0xed97f7562, 0x0}}, 0xc000c4c600, 0xc0043a1a38, 0xc0043a1a28, 0xc0043a1948, ...)
        /home/andrei/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1820 +0x511
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd(0xc000c4c600, {0x5208c88, 0xc001289d40})
        /home/andrei/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1822 +0x43a
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run(0xc000c4c600, {0x5208c88, 0xc001136ec0}, 0xc0012a14a0, {0xc00375de10, 0x4201fd, 0xc00420b220}, 0xc005881ee0)
        /home/andrei/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1676 +0x285
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn(0xc003ab2000, {0x5208c88, 0xc001136ec0}, {0xc004892300}, {0x3, 0xc001287ca0, 0xc00420b220}, 0xc000854048)
        /home/andrei/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:681 +0x9a
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).processCommandsAsync.func1()
        /home/andrei/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:669 +0x2f5
created by github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).processCommandsAsync
        /home/andrei/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:597 +0x273
```


interestingly, if I run a race test, I get a different scenario, but for the same span:


```
COCKROACH_DEBUG_SPAN_USE_AFTER_FINISH=true make testrace PKG=./pkg/ccl/streamingccl/streamproducer
```

```
panic: use of Span after Finish. Span: batch flow coordinator. Finish previously called at: goroutine 6725 [running]:
runtime/debug.Stack()
        /home/andrei/goroot/src/runtime/debug/stack.go:24 +0x72
github.com/cockroachdb/cockroach/pkg/util/tracing.(*Span).finishInternal(0xc002789b00)
        /home/andrei/src/github.com/cockroachdb/cockroach/pkg/util/tracing/span.go:256 +0xff
github.com/cockroachdb/cockroach/pkg/util/tracing.(*Span).Finish(0xc0048f4d50)
        /home/andrei/src/github.com/cockroachdb/cockroach/pkg/util/tracing/span.go:247 +0x2f
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*BatchFlowCoordinator).Run(0xc00284d340, {0x8ef49c8, 0xc00449c0c0})
        /home/andrei/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/flow_coordinator.go:287 +0x4f6
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*vectorizedFlow).Run(0xc001b6c180, {0x8ef49c8, 0xc00449c0c0}, 0x1)
        /home/andrei/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/vectorized_flow.go:259 +0x329
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run(0xc004447100, 0xc00303f1d0, 0xc00284d970, 0xc003385300, 0xc0045ace00, 0xc002330210, 0x0)
        /home/andrei/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:563 +0xf95
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRun(0x8ef4a70, {0x8ef4a70, 0xc002f4a750}, 0x1, 0xc00303f1d0, 0xc000e1d100, {{0x8ef6210, 0xc003384f80}, 0x0}, 0xc0045ace00)
        /home/andrei/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:1401 +0x24a
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithDistSQLEngine(0xc00232f900, {0x8ef4a70, 0xc002f4a750}, 0xc00232ff50, 0xc002f4a750, {0x7f38053493f8, 0xc000e1d128}, 0x0, 0xc002fd7978)
        /home/andrei/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1408 +0xcec
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).dispatchToExecutionEngine(0xc00232f900, {0x8ef4a70, 0xc002f4a750}, 0xc00232ff50, {0x7f38053493f8, 0xc000e1d128})
        /home/andrei/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1085 +0x12c5
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState(0xc00232f900, {0x8ef4a70, 0xc002f4a750}, {{0x8f22e38, 0xc0005efa20}, {0xc003a290bc, 0x27}, 0x0, 0x0}, 0xc0004e9000, ...)
        /home/andrei/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:722 +0x342b
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt(0xc00232f900, {0x8ef4a70, 0xc002f4a6f0}, {{0x8f22e38, 0xc0005efa20}, {0xc003a290bc, 0x27}, 0x0, 0x0}, 0xc0004e9000, ...)
        /home/andrei/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:135 +0x93d
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execPortal(0xc00232f900, {0x8ef4a70, 0xc002f4a6f0}, {0xc0004e9000, {0xc334738, 0x0, 0x0}, {0xc0031b270a, 0x3, 0x3}, ...}, ...)
        /home/andrei/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:222 +0x1dd
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd.func2({{0xc003a2910c, 0x0}, 0x0, {0x16c2802b, 0xed97f75fa, 0x0}}, 0xc00232f900, 0xc0048f7810, 0xc0048f7800, 0xc0047135d8, ...)
        /home/andrei/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1820 +0xa1d
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd(0xc00232f900, {0x8ef49c8, 0xc00177ff80})
        /home/andrei/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1822 +0x6f9
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run(0xc00232f900, {0x8ef49c8, 0xc000eddb80}, 0xc003173d68, {0x5f94a14, 0x0, 0xc0007701e0}, 0xc0028d72e0)
        /home/andrei/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1676 +0x5ff
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn(0xc00442dc00, {0x8ef49c8, 0xc000eddb80}, {0xc00442dc00}, {0x3, 0xc005a8cbb1, 0xc0007701e0}, 0xc00070a1b0)
        /home/andrei/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:681 +0x14c
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).processCommandsAsync.func1()
        /home/andrei/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:669 +0x406
created by github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).processCommandsAsync
        /home/andrei/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:597 +0x405


goroutine 6695 [running]:
github.com/cockroachdb/cockroach/pkg/util/tracing.(*Span).detectUseAfterFinish(0xc002789b00)
        /home/andrei/src/github.com/cockroachdb/cockroach/pkg/util/tracing/span.go:183 +0x245
github.com/cockroachdb/cockroach/pkg/util/tracing.(*Span).Tracer(...)
        /home/andrei/src/github.com/cockroachdb/cockroach/pkg/util/tracing/span.go:226
github.com/cockroachdb/cockroach/pkg/util/log.getSpanOrEventLog({0x8ef4a70, 0xc002f4ab70})
        /home/andrei/src/github.com/cockroachdb/cockroach/pkg/util/log/trace.go:92 +0x65
github.com/cockroachdb/cockroach/pkg/util/log.vEventf({0x8ef4a70, 0xc002f4ab70}, 0x0, 0x1, 0x303f0e18, {0x606798b, 0x36}, {0xc0046ac638, 0x1, 0x1})
        /home/andrei/src/github.com/cockroachdb/cockroach/pkg/util/log/trace.go:206 +0x125
github.com/cockroachdb/cockroach/pkg/util/log.VEventf(...)
        /home/andrei/src/github.com/cockroachdb/cockroach/pkg/util/log/trace.go:232
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLReceiver).SetError(0xc0045a0700, {0x8e63e00, 0xc00019c280})
        /home/andrei/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:912 +0x18d
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLReceiver).pushMeta(0xc0045a0700, 0xc000472540)
        /home/andrei/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:959 +0xa88
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLReceiver).Push(0xc0045a0700, {0x0, 0x0, 0x0}, 0xc000472540)
        /home/andrei/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:1047 +0x4d6
github.com/cockroachdb/cockroach/pkg/sql/rowflow.(*copyingRowReceiver).Push(0xc0045a0e00, {0x0, 0x0, 0x0}, 0x0)
        /home/andrei/src/github.com/cockroachdb/cockroach/pkg/sql/rowflow/row_based_flow.go:454 +0x2d6
github.com/cockroachdb/cockroach/pkg/sql/execinfra.Run({0x8ef4a70, 0xc005c36870}, {0x8f1b078, 0xc000294a80}, {0x8e860d8, 0xc0045a0e00})
        /home/andrei/src/github.com/cockroachdb/cockroach/pkg/sql/execinfra/base.go:189 +0x86
github.com/cockroachdb/cockroach/pkg/sql/execinfra.(*ProcessorBaseNoHelper).Run(0xc000294a80, {0x8ef49c8, 0xc005e1df40})
        /home/andrei/src/github.com/cockroachdb/cockroach/pkg/sql/execinfra/processorsbase.go:733 +0xeb
github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowBase).Run(0xc002497380, {0x8ef49c8, 0xc005e1df40}, 0x0)
        /home/andrei/src/github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow.go:453 +0x42a
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run(0xc004447100, 0xc002d90a50, 0x0, 0xc004f47998, 0xc0045a0700, 0xc003a08c00, 0x0)
        /home/andrei/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:563 +0xf95
github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/changefeeddist.StartDistChangefeed({0x8ef4a70, 0xc002f4ab70}, {0x8f7c338, 0xc00232ff50}, 0x0, {0x0, {0x0, 0x0}, 0xc005c36600, {0x16ccf2677d60748d, ...}}, ...)
        /home/andrei/src/github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/changefeeddist/distflow.go:144 +0xa28
github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamproducer.streamKVs({0x8ef4a70, 0xc002f4ab70}, {0x8fdcd60, 0xc00232ff50}, {0x5a3678, 0x5a3f2c, 0x0}, {0xc005c365d0, 0x1, 0x1}, ...)
        /home/andrei/src/github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamproducer/replication_stream_planning.go:105 +0x3fb
github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamproducer.doCreateReplicationStream({0x8ef4a70, 0xc002f4ab70}, {0x8fdcd60, 0xc00232ff50}, 0xc003d265d0, 0x0)
        /home/andrei/src/github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamproducer/replication_stream_planning.go:164 +0x7f5
github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamproducer.createReplicationStreamHook.func1({0x8ef4a70, 0xc002f4ab70}, {0x499537, 0x496761, 0xc00356c7d0}, 0xe2451e)
        /home/andrei/src/github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamproducer/replication_stream_planning.go:202 +0x6f
github.com/cockroachdb/cockroach/pkg/sql.(*hookFnNode).startExec.func1()
        /home/andrei/src/github.com/cockroachdb/cockroach/pkg/sql/planhook.go:148 +0xec
created by github.com/cockroachdb/cockroach/pkg/sql.(*hookFnNode).startExec
        /home/andrei/src/github.com/cockroachdb/cockroach/pkg/sql/planhook.go:147 +0x1d3
FAIL    github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamproducer    5.650s
```

--- a/pkg/sql/colflow/flow_coordinator.go
+++ b/pkg/sql/colflow/flow_coordinator.go
@@ -254,14 +254,6 @@ func (f *BatchFlowCoordinator) pushError(err error) execinfra.ConsumerStatus {
 // then shuts it down.
 func (f *BatchFlowCoordinator) Run(ctx context.Context) {
 	status := execinfra.NeedMoreRows
-	// Make sure that we close the coordinator and notify the batch receiver in
-	// all cases.
-	defer func() {
-		if err := f.close(); err != nil && status != execinfra.ConsumerClosed {
-			f.pushError(err)
-		}
-		f.output.ProducerDone()
-	}()
 
 	ctx, span := execinfra.ProcessorSpan(ctx, "batch flow coordinator")
 	if span != nil {
@@ -269,8 +261,20 @@ func (f *BatchFlowCoordinator) Run(ctx context.Context) {
 			span.SetTag(execinfrapb.FlowIDTagKey, attribute.StringValue(f.flowCtx.ID.String()))
 			span.SetTag(execinfrapb.ProcessorIDTagKey, attribute.IntValue(int(f.processorID)))
 		}
-		defer span.Finish()
 	}
+
+	// Make sure that we close the coordinator and notify the batch receiver in
+	// all cases.
+	defer func() {
+		if err := f.close(); err != nil && status != execinfra.ConsumerClosed {
+			f.pushError(err)
+		}
+		f.output.ProducerDone()
+		// Note that f.close is only safe to call before finishing the tracing
+		// span because some components might still use the span when they are
+		// being closed.
+		span.Finish()
+	}()
 
 	if err := f.init(ctx); err != nil {
 		f.pushError(err)

